### DISCUSSION TITLE
How to use DataTemplates in Style Selectors

### DISCUSSION THREAD

**Discussion by StefanKoell** [AvaloniaUI/Avalonia#13180] [Q&A] [answered]
Hi,

I'm trying to figure out how I can set two different visuals in a `MenuItem.Icon` depending on the type. For this I created the following style selectors:

```xaml
                    <Menu ItemsSource="{Binding MainMenuViewModel.MenuCommands}">
                        <Menu.Styles>
                            <Style Selector="MenuItem" x:DataType="commands:AppCommand">
                                <Setter Property="Header" Value="{Binding MenuCaption}" />
                                <Setter Property="Icon">
                                    <Template>
                                        <Border Background="Blue" Width="24" Height="24">
                                            <Border.DataTemplates>
                                                <DataTemplate x:DataType="commands:AppCommandChecked">
                                                    <Border Background="Red" Width="20" Height="20" />
                                                </DataTemplate>
                                                <DataTemplate x:DataType="commands:AppCommand">
                                                    <Border Background="Green" Width="20" Height="20" />
                                                </DataTemplate>
                                            </Border.DataTemplates>
                                        </Border>
                                    </Template>
                                </Setter>
                                <Setter Property="ItemsSource" Value="{Binding AppCommands}" />
                                <Setter Property="Command" Value="{Binding Command}" />
                            </Style>
                        </Menu.Styles>
                    </Menu>
```

All the bindings work great, except the one for the `Icon` property. I couldn't find any samples for this specific scenario. For testing/debugging I subscribed to the DataContextChanged event on the "outer" Blue border and I can see that the correct DataContext is set at that level. But then, the DataTemplates are not selected according to the specified DataType.

Is this a bug or am I doing something wrong?

Thanks,
Stefan

**Comment by timunie** [comment]
I'm also suprised that this doesn't work for you. Probably you need to implement IDataTemplate or use a FuncDataTemplate, as you have a better way to select DataTemplate on your own.

**Reply by StefanKoell** [reply]
I'm not sure I understand (still new to AvaloniaUI). I was wondering if this should work that way or if I'm not doing it right. If this should work, it's a bug and I guess it's in the interest for everyone to get that fixed, right?

**Reply by timunie** [reply]
Please file a minimal sample App for  me to investigate. If I can reproduce, I'll make an issue of it.

**Reply by StefanKoell** [reply]
Hi Tim!

Thanks for looking into it. I've created a minimal repro project which should get you the same result:
[MenuIcon.zip](https://github.com/AvaloniaUI/Avalonia/files/12849604/MenuIcon.zip)

Regards,
Stefan

**Reply by StefanKoell** [reply, marked as answer]
Answering my own question here. It turned out that I had to use a `ContentPresenter` (with `Content="{Binding}"`) and the appropriate data templates defined. This seems to work.

**Reply by timunie** [reply]
ContentControl instead of ContentPresnter may be safer outside ControlTemplates. If it works, fine. Just wanted to let you know it was not desined to work like this and thus may stop working.
